zebra/rt_netlink.c

```diff
@@ -3744,15 +3744,14 @@ static int netlink_ipneigh_change(struct nlmsghdr *h, int len, ns_id_t ns_id)
 	if (tb[NDA_LLADDR]) {
 		/* copy LLADDR information */
 		l2_len = RTA_PAYLOAD(tb[NDA_LLADDR]);
-		memcpy(&mac, RTA_DATA(tb[NDA_LLADDR]), l2_len);
 	}
 	if (l2_len == IPV4_MAX_BYTELEN || l2_len == 0) {
 		union sockunion link_layer_ipv4;
 
 		if (l2_len) {
 			sockunion_family(&link_layer_ipv4) = AF_INET;
 			memcpy((void *)sockunion_get_addr(&link_layer_ipv4),
-			       &mac, l2_len);
+			       RTA_DATA(tb[NDA_LLADDR]), l2_len);
```

Comment:
If the number of bytes is truly unknown doesn't this memcpy have the same problem

Reply:
imo this is wrong.  We want to copy the LLADDR into mac when it is 6.    This completely breaks that.

Reply:
No, here we're inside the following condition: `l2_len == IPV4_MAX_BYTELEN || l2_len == 0`. And also it's under `if (l2_len)`, so we know it's exactly 4 bytes here.

Reply:
@donaldsharp please look at the later code. This `mac` variable is re-initialized to zeroes on line 3791 anyway, and set to the correct value on line 3810 only when the length is 6 bytes. The existing code already takes care of everything needed.

Reply:
this fixes the crash, but we dont explicitly say that l2_len other than 4 or 6 bytes should be ignored.
like it is done, at line 3819, we consider NDA_LLADDR is a 4 byte length.
```memcpy(&mac, RTA_DATA(tb[NDA_LLADDR]), ETH_ALEN);```
I think somewhere at line 3765 , we should add a control with length:
```
if (h->nlmsg_type == RTM_GETNEIGH || l2_len != ETH_ALEN)
   return 0;
```

Reply:
No, we should not do that, because `l2_len == 0` is totally fine for `RTM_DELNEIGH` request.
`RTM_NEWNEIGH` already has all necessary checks for the length.